apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,616 @@
+  if (!(/^(\d+)\.?(\d*)(e-\d+)?$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an floating-point number, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is in a list of possibilities (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertInChoices(
+  tag: string,
+  key: string,
+  val: string,
+  choices: string[],
+  errors: string[],
+) {
+  if (!(choices.includes(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be in ${choices}, but value provided is ${val}`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is a boolean (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertBool(tag: string, key: string, val: string, errors: string[]) {
+  assertInChoices(tag, key, val, ['true', 'false'], errors);
+}
+
+/**
+ * Checks that a tag has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkTag(ast: any): string[] {
+  switch (ast.tagName) {
+    case 'pl-multiple-choice':
+      return checkMC(ast);
+    case 'pl-integer-input':
+      return checkInt(ast);
+    case 'pl-number-input':
+      return checkNum(ast);
+    case 'pl-answer':
+      return checkMCAnswer(ast);
+    case 'pl-overlay':
+      return checkOverlay(ast);
+    case 'pl-location':
+      return checkLocation(ast);
+    case 'pl-order-blocks':
+      return checkOrderBlocks(ast);
+    case 'pl-matrix-input':
+      return checkMatrix(ast);
+  }
+  return [];
+}
+
+/**
+ * Checks that a multiple choice question tag has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkMC(ast: any): string[] {
+  const errors: string[] = [];
+  let usedAnswersName = false;
+  let displayDropdown = false;
+  let usedAllOfTheAbove = false;
+  let usedNoneOfTheAbove = false;
+  let usedAllOfTheAboveFeedback = false;
+  let usedNoneOfTheAboveFeedback = false;
+  let usedSize = false;
+  const optionsOfTheAbove = ['false', 'random', 'correct', 'incorrect'];
+  for (const attr of ast.attrs) {
+    const key = attr.name;
+    const val = attr.value;
+    switch (key) {
+      case 'answers-name':
+        usedAnswersName = true;
+        break;
```

Comment:
Using a loop and switch-case for this feels a bit weird and overly verbose. Something like this would probably be more idiomatic:

```ts
const hasAnswersName = ast.attrs.includes(a => a.name === 'answers-name');
```

You could even make a function out of that:

```ts
const hasAnswersName = hasAttribute(ast.attrs, 'answers-name');
```

As I'm writing this out, I imagine the loop + case is here so that you can handle unexpected attributes in the `default:` case. I'd be inclined to more closely mirror the way that element code itself checks all this via something like this `check_attribs` function:

https://github.com/PrairieLearn/PrairieLearn/blob/ab0010dc0a461a47f78f1e33412e8a90964a4b1e/apps/prairielearn/elements/pl-multiple-choice/pl-multiple-choice.py#L400

I won't block on this if you feel that the way you've written things is the best fit for now.